Replace linked list based shadows with one based on FixedVector

weinig opened the pull request:
#### e0985bdd8d81bdb4a74920c115e88dc1e5f57ca4
<pre>
Replace linked list based shadows with one based on FixedVector
<a href="https://bugs.webkit.org/show_bug.cgi?id=293749">https://bugs.webkit.org/show_bug.cgi?id=293749</a>

Reviewed by Darin Adler.

Switches the data structure used for storing box shadows and text shadows
for RenderStyle from an intrusive linked list to FixedVector. The new
implementation is equally efficient for shadow lists of zero or one, and
more efficient for anything more as the intrusive pointer is eliminated.

* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/TrailingArray.h:
* Tools/TestWebKitAPI/Tests/WTF/FixedVector.cpp:
    - Rename FixedVector::createWithSizeFromGenerator to FixedVector::createWithSizeFromFailableGenerator
      and add a new createWithSizeFromGenerator that doesn&apos;t allow failure cases.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
    - Adopt new FixedVector::createWithSizeFromGenerator by removing now unnecessary std::optional return.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/style/values/borders/StyleShadow.h: Added.
    - Add StyleShadow.h for algorithms shared between Style::BoxShadow and Style::TextShadow.

* Source/WebCore/accessibility/AXObjectCache.cpp:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
* Source/WebCore/bindings/js/JSDOMConvertVariadic.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/BackgroundPainter.h:
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.h:
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
* Source/WebCore/rendering/LegacyInlineFlowBox.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/StyledMarkedText.cpp:
* Source/WebCore/rendering/StyledMarkedText.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
* Source/WebCore/rendering/TextBoxPainter.h:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.h:
* Source/WebCore/rendering/TextPaintStyle.cpp:
* Source/WebCore/rendering/TextPaintStyle.h:
* Source/WebCore/rendering/TextPainter.cpp:
* Source/WebCore/rendering/TextPainter.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.cpp:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/ShadowData.cpp: Removed.
* Source/WebCore/rendering/style/ShadowData.h: Removed.
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/borders/StyleBoxShadow.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h:

Canonical link: <a href="https://commits.webkit.org/295646@main">https://commits.webkit.org/295646@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebe28dda1cbd8b16bc3d5d6eda114fc902adce99

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105749 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25500 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110946 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56345 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107790 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25987 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80309 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108755 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20330 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95426 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60620 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20046 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13523 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55784 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98390 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89774 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13562 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113795 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104368 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32889 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24238 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89387 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33253 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91657 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89056 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22700 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33927 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11731 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28349 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32814 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38224 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/128679 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32560 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35125 "Found 2 new JSC stress test failures: wasm.yaml/wasm/gc/ref-i31-eq.js.default-wasm, wasm.yaml/wasm/gc/ref-i31-eq.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35909 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34158 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2f318ddc4373eca6c6afa48c3a7fc9d13cedf463. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46061

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0bf43940602900e46cb8b878154c85a9581422a1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46061

weinig commented on Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
```diff
@@ -622,11 +623,18 @@ void SVGTextBoxPainter<TextBoxPath>::paintTextWithShadows(const RenderStyle& sty
 
-        shadow = shadow->next();
-    } while (shadow);
+    if (shadows.isEmpty()) {
+        draw(nullptr, false);
+        return;
+    }
```
I don’t think we can.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4ef4b2d81c408aa5d7366f555ee524144a73adc0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46061

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ebe28dda1cbd8b16bc3d5d6eda114fc902adce99. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46061

weinig committed:
Replace linked list based shadows with one based on FixedVector
https://bugs.webkit.org/show_bug.cgi?id=293749

Reviewed by Darin Adler.

Switches the data structure used for storing box shadows and text shadows
for RenderStyle from an intrusive linked list to FixedVector. The new
implementation is equally efficient for shadow lists of zero or one, and
more efficient for anything more as the intrusive pointer is eliminated.

* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/TrailingArray.h:
* Tools/TestWebKitAPI/Tests/WTF/FixedVector.cpp:
    - Rename FixedVector::createWithSizeFromGenerator to FixedVector::createWithSizeFromFailableGenerator
      and add a new createWithSizeFromGenerator that doesn't allow failure cases.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
    - Adopt new FixedVector::createWithSizeFromGenerator by removing now unnecessary std::optional return.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/style/values/borders/StyleShadow.h: Added.
    - Add StyleShadow.h for algorithms shared between Style::BoxShadow and Style::TextShadow.

* Source/WebCore/accessibility/AXObjectCache.cpp:
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
* Source/WebCore/bindings/js/JSDOMConvertVariadic.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/BackgroundPainter.h:
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.h:
* Source/WebCore/rendering/LegacyInlineFlowBox.cpp:
* Source/WebCore/rendering/LegacyInlineFlowBox.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/StyledMarkedText.cpp:
* Source/WebCore/rendering/StyledMarkedText.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
* Source/WebCore/rendering/TextBoxPainter.h:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.h:
* Source/WebCore/rendering/TextPaintStyle.cpp:
* Source/WebCore/rendering/TextPaintStyle.h:
* Source/WebCore/rendering/TextPainter.cpp:
* Source/WebCore/rendering/TextPainter.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.cpp:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/ShadowData.cpp: Removed.
* Source/WebCore/rendering/style/ShadowData.h: Removed.
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/borders/StyleBoxShadow.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.cpp:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h:

Canonical link: https://commits.webkit.org/295646@main
